expirationd.lua

```diff
@@ -327,48 +390,202 @@ end
---     * iteration_delay       -- Max sleep time between iterations (in seconds).
---     * full_scan_delay       -- Sleep time between full scans (in seconds).
---     * force                 -- Run task even on replica.
---  }
+--- Run a scheduled task to check and process (expire) tuples in a given space.
+--
+-- How expirationd works in general:
+--
+-- 1. Process min(`space_length`, `tuples_per_iteration`) tuples at once.
+--
+-- 2. Sleep `tuples_per_iteration` × `full_scan_time` / `space_length` (but not
+--    beyond 1 second).
+--
+-- 3. Repeat 1-2 until the whole space will be traversed.
+--
+-- 4. Sleep 1 second.
+--
+-- 5. Repeat 1-4.
+--
+--
+-- @string name
+--     Task name.
+-- @string space_id
+--     Space to look in for expired tuples.
+-- @func is_tuple_expired
+--     Function, must accept tuple and return `true` or `false` (is tuple
+--     expired or not), receives `args` and `tuple` as arguments.
+--
+-- Example of function:
+--
+-- ```
+-- local function is_tuple_expired(args, tuple)
+--     local tuple_expire_time = get_field(tuple, args.field_no)
+--     if type(tuple_expire_time) ~= 'number' then
+--         return true
+--     end
+--     local current_time = fiber.time()
+--
+--     return current_time >= tuple_expire_time
+-- end
+-- ```
+--
+-- @table[opt] options
+--     Table with named options.
+-- @table[opt] options.args
+--     Passed to `is_tuple_expired()` and `process_expired_tuple()` as
+--     an additional context.
+-- @boolean[opt] options.atomic_iteration
+--     False (default) to process each tuple as a single transaction and true
+--     to process tuples from each batch in a single transaction.
+-- @boolean[opt] options.force
+--     expirationd doesn't work on replica by default. With option `force` it
+--     should be possible to force work on replica.
+-- @number[opt] options.full_scan_delay
+--     Sleep time between full scans (in seconds). It is allowed to pass an FFI
+--     number: `1LL`, `1UL` etc. Default value is 1 sec.
+-- @number[opt] options.full_scan_time
+--     Time required for a full index scan (in seconds). It is allowed to pass
+--     an FFI number: `1LL`, `1UL` etc. `full_scan_time` used for calculation
+--     of time during which fiber sleeps between iterations. Default value is
+--     3600.
+-- @string[opt] options.index
+--     Name or id of the index to iterate on. If omitted, will use the primary
+--     index. If there's no index with this name, will throw an error.
+--     Supported index types are TREE and HASH, using other types will result
+--     in an error.
+-- @func[opt] options.iterate_with
+--     Function which returns an iterator object which provides tuples to
+--     check, considering the `start_key`, `process_while` and other options.
+--     There's a default function which can be overriden with this parameter.
+--
+-- Example of function:
+--
+-- ```
+-- local function iterate_with()
+--     return true
+-- end
```

Comment:
Just above:

> Function, which returns an iterator object \<...\>

But this function does not return an iterator object (`gen`, `param` `state` triplet, where `gen` is callable).

Reply:
And I don't see any information about arguments, which is passed to the function.

Reply:
Removed example and instead described default function:
```lua
@@ -463,18 +460,25 @@ end
 -- @func[opt] options.iterate_with
 --     Function which returns an iterator object which provides tuples to
 --     check, considering the `start_key`, `process_while` and other options.
---     There's a default function which can be overriden with this parameter.
---
--- Example of function:
+--     When option is nil default function is used. Default function returns
+--     iterator returned by [index_object:pairs()][1], where `index` is a
+--     primary index or index that specified with argument `options.index`:
 --
 -- ```
--- local function iterate_with()
---     return true
--- end
+--  index:pairs(option.start_key(), {
+--     iterator = task.iterator_type
+--  }):take_while(
+--         function()
+--             return option.process_while()
+--         end
+--    )
 -- ```
+--
+--     [1]: https://www.tarantool.io/en/doc/latest/reference/reference_lua/box_index/pairs/.
+--
```

> And I don't see any information about arguments, which is passed to the function.

it is a question how to describe it, because `task` is an internal object. Perhaps we need to change arguments passed to `iterate_with` to make possible to describe it in documentation. What do you think?

Reply:
Okay.